### PR TITLE
R2 final review

### PR DESCRIPTION
**Pull request by GoodluckH**

### CODE REVIEW

**Comment by GoodluckH** [comment]
[30d420e](https://github.com/Arrow-air/svc-assets/pull/16/commits/30d420eaee77a795c990613beb59edd69097a4e3) closes #17

File: README.md [hunk 1]
```diff
@@ -1,24 +1,23 @@
...
 
-*TODO after cloning:*
-
-1. *Replace the repository name for each:*
-
-![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/Arrow-air/svc-template?include_prereleases)
-![Sanity Checks](https://github.com/arrow-air/svc-template-rust/actions/workflows/sanity_checks.yml/badge.svg?branch=main)
-![Rust Checks](https://github.com/arrow-air/svc-template-rust/actions/workflows/rust_ci.yml/badge.svg?branch=main)
-![Python Flake8](https://github.com/arrow-air/svc-template-rust/actions/workflows/python_ci.yml/badge.svg?branch=main)
+![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/Arrow-air/svc-assets?include_prereleases)
```
Comment (amsmith-pro) [inline README.md:7]: There's another badge for stable releases, see `svc-storage`

File: server/src/main.rs [hunk 2]
```diff
@@ -1,21 +1,65 @@
...
+pub mod svc_assets {
     #![allow(unused_qualifications, missing_docs)]
     include!("grpc.rs");
 }
 
-use svc_template_rust::template_rust_rpc_server::{TemplateRustRpc, TemplateRustRpcServer};
-use svc_template_rust::{QueryIsReady, ReadyResponse};
-use tonic::{transport::Server, Request, Response, Status};
+mod rest_api;
+//TODO - replace this with database calls when ready
```
Comment (amsmith-pro) [inline server/src/main.rs]: is this still applicable?

File: server/src/rest_api.rs [hunk 3]
```diff
@@ -0,0 +1,1371 @@
...
+) -> Result<Json<Vec<Aircraft>>, (StatusCode, String)> {
+    let filter = AdvancedSearchFilter::search_is_not_null(String::from("deleted_at"));
+    let vehicle_client_option = grpc_clients.storage_vehicle.get_client().await;
+    if vehicle_client_option.is_none() {
+        let error_msg = "svc-storage unavailable.".to_string();
+        req_error!("(get_all_aircraft) {}", &error_msg);
+        return Err((StatusCode::SERVICE_UNAVAILABLE, error_msg));
+    }
+
+    let mut vehicle_client = vehicle_client_option.unwrap();
```
Comment (amsmith-pro) [inline server/src/rest_api.rs]: handle unwrap

File: server/src/rest_api.rs [hunk 4]
```diff
@@ -0,0 +1,1371 @@
...
+
+    let vehicle_response = vehicle_client.search(filter.clone()).await;
+
+    if vehicle_response.is_err() {
+        let error_msg = "svc-storage unavailable.".to_string();
+        req_error!("(get_all_aircraft) {}", &error_msg);
+        return Err((StatusCode::SERVICE_UNAVAILABLE, error_msg));
+    }
+    let mut vehicles = vehicle_response.unwrap().into_inner().list;
```
Comment (amsmith-pro) [inline server/src/rest_api.rs]: handle unwrap

last comment about unwraps, I will check for them again on re-review

File: server/src/rest_api.rs [hunk 5]
```diff
@@ -0,0 +1,1371 @@
...
+    ),
+    params(
+        ("id" = String, Path, description = "Operator id"),
+    )
+)]
+pub async fn get_operator(
+    Extension(mut _grpc_clients): Extension<GrpcClients>,
+    Path(operator_id): Path<String>,
+) -> Result<Json<Operator>, (StatusCode, String)> {
+    req_debug!("get_operator({})", operator_id);
```
Comment (amsmith-pro) [inline server/src/rest_api.rs]: I like this, please add a debug statement to the beginning of the other api handlers like you have here

**Comment by amsmith-pro** [comment]
## Documents
- [x] ICD
  - [x] Current interfaces (grpc & rest)
  - [x] Current relevant document links
- [x] SDD
  - [x] Description of grpc & rest handler implementations
  - [x] Current sequence diagrams where applicable
  - [x] Current relevant document links
- [x] CONOPS
  - [x] Current information
  - [x] Bibliography (if any citations)
- [x] README (see svc-cargo for example)
  - [x] Badges
  - [x] Accurate installation instructions
  - [x] Current links to other documents
  - [x] Arrow DAO links

## Code (not applicable to examples/ folder)
- [x] No unhandled unwrap()s
- [x] Accurate (!) code comments on all functions
- [x] No `# Arguments` and `# Returns` in function comments
- [x] Function arguments checked where possible
- [x] Sufficient logger (info, warn, and debug) messages
- [x] Sensible Types (unsigned vs. signed)

## REST API
- [x] Return codes accurately documented with openapi
- [x] Return codes make sense
- [x] Sufficiently descriptive error messages

## Cargo.toml(s)
- [x] Package versions correct
- [x] No minor versions specified
- [ ] Links to crates.io package, not git branch

File: server/src/rest_api.rs [hunk 6]
```diff
@@ -0,0 +1,1371 @@
...
+    if !is_uuid(&operator_id) {
+        return Err((StatusCode::BAD_REQUEST, "Invalid operator id".to_string()));
+    }
+    // Get Client
+    // TODO let _client_option = grpc_clients.storage.get_client().await;
```
Comment (amsmith-pro) [inline server/src/rest_api.rs]: Update TODOs with estimated release i.e. `TODO R3`, most likely R3 for many of these

File: proto/svc-assets-grpc.proto [hunk 7]
```diff
@@ -3,7 +3,7 @@ syntax = "proto3";
 package grpc;
 
 // Heartbeat
-service TemplateRustRpc {
+service SvcAssetsRpc {
```
Comment (amsmith-pro) [inline proto/svc-assets-grpc.proto]: Call this RpcService, we're going to make this the same across all services

File: server/Cargo.toml [hunk 8]
```diff
@@ -17,20 +18,34 @@ categories = [
...
+tokio         = { version = "1.20", features = ["full"] }
+tokio-util    = "0.7"
+tonic         = "0.8"
+tonic-health  = "0.7"
+utoipa        = { version = "2.1", features = ["axum_extras", "chrono"] }
+uuid          = { version = "1.2", features = ["v4", "serde"] }
+
+[dependencies.svc-storage-client-grpc]
+git = "https://github.com/Arrow-air/svc-storage"
+tag = "v0.9.0-develop.15"
```
Comment (amsmith-pro) [inline server/Cargo.toml]: this can be v0.10.0 now

File: server/Cargo.toml [hunk 9]
```diff
@@ -1,36 +1,50 @@
 [package]
-description = "Arrow template service GRPC server"
-edition     = "2021"                                                       # Allowed values are 2015, 2018, 2021
-homepage    = "https://www.arrowair.com/docs/documentation/services/intro"
-keywords    = ["vtol", "server", "template"]                               # max 5
-license     = "BUSL-1.1"
-name        = "svc-template-rust"
-repository  = "https://github.com/Arrow-air/svc-template-rust"
-version     = "0.2.0-develop.2"
+authors     = ["GoodluckH"]
```
Comment (amsmith-pro) [inline server/Cargo.toml]: can move authors to toplevel Cargo.toml

File: server/src/grpc/server.rs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+#[cfg(not(tarpaulin_include))]
+pub async fn grpc_server(config: Config) {
+    grpc_debug!("(grpc_server) entry.");
+
+    // GRPC Server
+    let grpc_port = config.docker_port_grpc;
+    let full_grpc_addr: SocketAddr = match format!("[::]:{}", grpc_port).parse() {
+        Ok(addr) => addr,
+        Err(e) => {
+            grpc_error!("Failed to parse gRPC address: {}", e);
```
Comment (amsmith-pro) [inline server/src/grpc/server.rs]: precede with (grpc_server) to keep the format

Comment (amsmith-pro) [reply server/src/grpc/server.rs]: same with other calls in this function

File: server/src/main.rs [hunk 11]
```diff
@@ -1,56 +1,45 @@
...
+    dotenv().ok();
+    {
+        let log_cfg: &str = config.log_config.as_str();
+        if let Err(e) = log4rs::init_file(log_cfg, Default::default()) {
+            panic!("(logger) could not parse {}. {}", log_cfg, e);
+        }
+    }
 
-    let full_grpc_addr = format!("[::]:{}", grpc_port).parse()?;
+    // --------------------------------------------------
```
Comment (amsmith-pro) [inline server/src/main.rs]: Remove the START and END REST SECTION comments

File: server/src/rest/api.rs [hunk 12]
```diff
@@ -0,0 +1,1370 @@
...
+    responses(
+        (status = 200, description = "Assets successfully found", body = [Vertiport]),
+        (status = 503, description = "Could not connect to other microservice dependencies")
+    ),
+)]
+/// Get all vertiports from the database.
+pub async fn get_all_vertiports(
+    Extension(mut grpc_clients): Extension<GrpcClients>,
+) -> Result<Json<Vec<Vertiport>>, (StatusCode, String)> {
+    let filter = AdvancedSearchFilter::search_is_not_null(String::from("deleted_at"));
```
Comment (amsmith-pro) [inline server/src/rest/api.rs:207]: entry info!

Comment (amsmith-pro) [reply server/src/rest/api.rs:207]: check for the other api functions as well

File: server/src/rest_api.rs (hunk 5 above)
Comment (amsmith-pro) [reply server/src/rest_api.rs]: according to owlot it should be info!

**Review by amsmith-pro — CHANGES_REQUESTED** [review]

File: server/src/main.rs (hunk 2 above)
Comment (GoodluckH) [reply server/src/main.rs]: This comment is outdated. Please resolve.

File: server/src/rest/api.rs (hunk 12 above)
Comment (GoodluckH) [reply server/src/rest/api.rs:207]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: server/src/main.rs (hunk 11 above)
Comment (GoodluckH) [reply server/src/main.rs]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: server/src/grpc/server.rs (hunk 10 above)
Comment (GoodluckH) [reply server/src/grpc/server.rs]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: server/Cargo.toml (hunk 9 above)
Comment (GoodluckH) [reply server/Cargo.toml]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: server/Cargo.toml (hunk 8 above)
Comment (GoodluckH) [reply server/Cargo.toml]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: server/src/rest_api.rs (hunk 5 above)
Comment (GoodluckH) [reply server/src/rest_api.rs]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: server/src/rest_api.rs (hunk 4 above)
Comment (GoodluckH) [reply server/src/rest_api.rs]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: server/src/rest_api.rs (hunk 3 above)
Comment (GoodluckH) [reply server/src/rest_api.rs]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: README.md (hunk 1 above)
Comment (GoodluckH) [reply README.md:7]: [4428782](https://github.com/Arrow-air/svc-assets/pull/16/commits/4428782e670f74e05fb753342bf5365399d11255)

File: server/src/rest/api.rs [hunk 13]
```diff
@@ -966,11 +1085,12 @@ pub async fn update_aircraft(
                 schedule: payload.schedule.unwrap_or(vehicle.schedule),
                 last_maintenance: if let Some(last_maintenance) = payload.last_maintenance {
                     if let Some(last_maintenance) = last_maintenance {
-                        let time_stamp_result = Timestamp::from_str(&last_maintenance);
-                        if time_stamp_result.is_ok() {
-                            Some(time_stamp_result.unwrap())
-                        } else {
-                            None
+                        match Timestamp::from_str(&last_maintenance) {
```
Comment (amsmith-pro) [inline server/src/rest/api.rs:1088]: this is a little messy (too many nests) can you pull this block out of the `match` statement (above it)?

Same with the one below, then this PR is good to go

**Review by amsmith-pro — CHANGES_REQUESTED** [review]

File: server/src/rest/api.rs (hunk 13 above)
Comment (GoodluckH) [reply server/src/rest/api.rs:1088]: [75a9e0e](https://github.com/Arrow-air/svc-assets/pull/16/commits/75a9e0e9ad3ffef9f27e3b44da1fda4e9f61bda8)

**Review by amsmith-pro — APPROVED** [review]

**Pull request merged**